src/react-components/room/useMicrophone.js

```diff
@@ -23,29 +31,98 @@ export function useMicrophone(scene, updateRate = 50) {
+        setMicVolume(prevVolume => Math.max(0.15, Math.abs(prevVolume - nextVolume) > 0.05 ? nextVolume : prevVolume));
         timeout = setTimeout(updateMicVolume, updateRate);
       };
 
       updateMicVolume();
 
-      const onMicStateChanged = ({ enabled }) => {
-        setIsMuted(!enabled);
+      const onMicMutedStateChanged = ({ enabled }) => {
+        setIsMicMuted(!enabled);
+      };
+      APP.dialog.on("mic-state-changed", onMicMutedStateChanged);
+
+      const onMicEnabled = () => {
+        setIsMicEnabled(true);
+        setSelectedMicDeviceId(mediaDevicesManager.selectedMicDeviceId);
+        setMicDevices(mediaDevicesManager.micDevices);
+      };
+      const onMicDisabled = () => {
+        setIsMicEnabled(false);
+        setSelectedMicDeviceId(mediaDevicesManager.selectedMicDeviceId);
+        setMicDevices(mediaDevicesManager.micDevices);
+      };
+      scene.addEventListener(MediaDevicesEvents.MIC_SHARE_ENDED, onMicDisabled);
+      scene.addEventListener(MediaDevicesEvents.MIC_SHARE_STARTED, onMicEnabled);
+
+      const onPermissionsChanged = ({ mediaDevice, status }) => {
+        if (mediaDevice === MediaDevices.MICROPHONE) {
+          setPermissionsStatus(status);
+          setSelectedMicDeviceId(mediaDevicesManager.selectedMicDeviceId);
+          setMicDevices(mediaDevicesManager.micDevices);
+        }
+      };
+      mediaDevicesManager.on(MediaDevicesEvents.PERMISSIONS_STATUS_CHANGED, onPermissionsChanged);
+
+      const onDeviceChange = () => {
+        setSelectedMicDeviceId(mediaDevicesManager.selectedMicDeviceId);
+        setMicDevices(mediaDevicesManager.micDevices);
       };
+      mediaDevicesManager.on(MediaDevicesEvents.DEVICE_CHANGE, onDeviceChange);
 
-      APP.dialog.on("mic-state-changed", onMicStateChanged);
+      setSelectedMicDeviceId(mediaDevicesManager.selectedMicDeviceId);
+      setMicDevices(mediaDevicesManager.micDevices);
 
       return () => {
         clearTimeout(timeout);
-        APP.dialog.off("mic-state-changed", onMicStateChanged);
+        APP.dialog.off("mic-state-changed", onMicMutedStateChanged);
+        scene.removeEventListener(MediaDevicesEvents.MIC_SHARE_ENDED, onMicDisabled);
+        scene.removeEventListener(MediaDevicesEvents.MIC_SHARE_STARTED, onMicEnabled);
+        mediaDevicesManager.off(MediaDevicesEvents.PERMISSIONS_STATUS_CHANGED, onPermissionsChanged);
+        mediaDevicesManager.off(MediaDevicesEvents.DEVICE_CHANGE, onDeviceChange);
       };
     },
-    [setVolume, scene, updateRate]
+    [
+      setIsMicMuted,
+      setMicVolume,
+      setIsMicEnabled,
+      setSelectedMicDeviceId,
+      setMicDevices,
+      setPermissionsStatus,
+      scene,
+      updateRate,
+      mediaDevicesManager
+    ]
   );
 
-  const toggleMute = useCallback(() => {
-    APP.dialog.toggleMicrophone();
-  }, []);
+  const toggleMute = useCallback(
+    () => {
+      if (mediaDevicesManager.isMicShared) {
```

Comment:
This reads pretty oddly. If its correct it feels a bit roundabout to be calling something on dialog based on state on mediaDevicesManager (which presumably would end up toggling the state on mediaDevicesManager)

Reply:
This makes sense too. I have made the devices manager the entry point for mic state changes. Make sense as it's already handling all the media devices state.